File: 00_ServerLibrary/RingBuffer.cpp
#include "stdafx.h"
#include "RingBuffer.h"

netlib::RingBuffer::RingBuffer()
{
	_bufferSize = 10000;
	_allocSize = _bufferSize + 1;
	_front = 0;
	_rear = 0;

	_buffer = new BYTE[_allocSize];
	::memset(_buffer, 0, _allocSize);
}

netlib::RingBuffer::RingBuffer(int32 bufferSize)
{
	_bufferSize = bufferSize;
	_allocSize = _bufferSize + 1;
	_front = 0;
	_rear = 0;

	_buffer = new BYTE[_allocSize];
	::memset(_buffer, 0, _allocSize);
}

netlib::RingBuffer::~RingBuffer()
{
	delete _buffer;
}

void netlib::RingBuffer::Resize(int32 size)
{
	int32 useSize = GetUseSize();
	BYTE* newBuffer = new BYTE[size];
	::memcpy(newBuffer, &_buffer[_front], useSize);
	delete _buffer;
	_buffer = newBuffer;
	_allocSize = size + 1;
	_bufferSize = size;
	_front = 0;
	_rear = useSize;
}

int32 netlib::RingBuffer::GetBufferSize()
{
	return _bufferSize;
}

int32 netlib::RingBuffer::GetUseSize()
{
	int32 rear = _rear;
	int32 front = _front;

	if (rear >= front)
	{
		return rear - front;
	}
	else
	{
		return _bufferSize + rear - front + 1;
	}
}

int32 netlib::RingBuffer::GetFreeSize()
{
	int32 rear = _rear;
	int32 front = _front;

	if (rear >= front)
	{
		return _bufferSize - rear + front;
	}
	else
	{
		return front - rear - 1;
	}
}

int32 netlib::RingBuffer::Enqueue(BYTE* data, int size)
{
	if (size > GetFreeSize())
	{
		return -1;
	}

	if (_rear + size > _bufferSize)
	{
		int32 chunkSize = _allocSize - _rear;
		::memcpy(&_buffer[_rear], data, chunkSize);
		data += chunkSize;
		::memcpy(&_buffer[0], data, size - chunkSize);
	}
	else
	{
		::memcpy(&_buffer[_rear], data, size);
	}

	_rear = (_rear + size) % _allocSize;

	return size;
}

int32 netlib::RingBuffer::Dequeue(BYTE* dest, int size)
{
	if (size > GetUseSize())
		return -1;

	if (_front + size > _bufferSize)
	{
		int32 chunkSize = _allocSize - _front;
		::memcpy(dest, &_buffer[_front], chunkSize);
		dest += chunkSize;
		::memcpy(dest, &_buffer[0], size - chunkSize);
	}
	else
	{
		::memcpy(dest, &_buffer[_front], size);
	}

	_front = (_front + size) % _allocSize;


	return size;
}

int32 netlib::RingBuffer::Peek(BYTE* dest, int size)
{
	if (size > GetUseSize())
		return -1;

	if (_front + size > _bufferSize)
	{
		int32 chunkSize = _allocSize - _front;
		::memcpy(dest, &_buffer[_front], chunkSize);
		dest += chunkSize;
		::memcpy(dest, &_buffer[0], size - chunkSize);
	}
	else
	{
		::memcpy(dest, &_buffer[_front], size);
	}


	return size;
}

int32 netlib::RingBuffer::DirectEnqueueSize()
{

	int32 rear = _rear;
	int32 front = _front;

	if (rear >= front)
	{
		if (front == 0)
		{
			return _bufferSize - rear;
		}
		else
		{
			return _bufferSize - rear + 1;
		}
	}
	else
	{
		return front - rear - 1;
	}
}

int32 netlib::RingBuffer::DirectDequeueSize()
{
	int32 rear = _rear;
	int32 front = _front;

	if (rear >= front)
	{
		return rear - front;
	}
	else
	{
		return _allocSize - front;
	}
}

int32 netlib::RingBuffer::MoveRear(int32 size)
{
	_rear = (_rear + size) % _allocSize;
	return size;
}

int32 netlib::RingBuffer::MoveFront(int32 size)
{
	_front = (_front + size) % _allocSize;
	return size;
}

void netlib::RingBuffer::Clear()
{
	_front = 0;
	_rear = 0;
}

File: 04_GameClient/Network/RingBuffer.cs
using System;
using System.Collections.Generic;
using Unity.VisualScripting;
using UnityEngine;

namespace Controller.Network
{
    public class RingBuffer
    {
        private Byte[] _buffer;
        public Int32 bufferSize { get; private set; }
        public Int32 allocSize { get; private set; }
        private Int32 _front;
        private Int32 _rear;

        public RingBuffer()
        {
            bufferSize = 10000;
            allocSize = bufferSize + 1;
            _front = 0;
            _rear = 0;
            _buffer = new byte[allocSize];
        }

        public void MoveFront(Int32 size)
        {
            _front = (_front + size) % allocSize;
        }
        
        public void MoveRear(Int32 size)
        {
            _rear = (_rear + size) % allocSize;
        }
        
        public Int32 GetUseSize()
        {
            Int32 nowRear = _rear;
            Int32 nowFront = _front;

            if (nowRear >= nowFront)
            {
                return nowRear - nowFront;
            }
            else
            {
                return bufferSize + nowRear - nowFront + 1;
            }
        }

        public Int32 GetFreeSize()
        {
            Int32 nowRear = _rear;
            Int32 nowFront = _front;

            if (nowRear >= nowFront)
            {
                return bufferSize - nowRear + nowFront;
            }
            else
            {
                return nowFront - nowRear - 1;
            }
        }

        public Int32 DirectEnqueueSize()
        {
            Int32 nowRear = _rear;
            Int32 nowFront = _front;

            if (nowRear >= nowFront)
            {
                if (nowFront == 0)
                {
                    return bufferSize - nowRear;
                }
                else
                {
                    return bufferSize - nowRear + 1;
                }
            }
            else
            {
                return nowFront - nowRear - 1;
            }
        }

        public Int32 DirectDequeueSize()
        {
            Int32 nowRear = _rear;
            Int32 nowFront = _front;

            if (nowRear >= nowFront)
            {
                return nowRear - nowFront;
            }
            else
            {
                return allocSize - nowFront;
            }
        }

        public Int32 Enqueue(ArraySegment<Byte> data, Int32 size)
        {
            if (size > GetFreeSize())
            {
                return -1;
            }

            if (_rear + size > bufferSize)
            {
                int chunkSize = allocSize - _rear;
                Array.Copy(data.Array, 0, _buffer, _rear, chunkSize);
                Array.Copy(data.Array, chunkSize, _buffer, 0, size - chunkSize);
            }
            else
            {
                Array.Copy(data.Array, 0, _buffer, _rear, size);
            }

            _rear = (_rear + size) % allocSize;

            return size;
        }

        public Int32 Dequeue(ArraySegment<Byte> dest, Int32 size)
        {
            if (size > GetUseSize())
            {
                return -1;
            }
            
            if (_front + size > bufferSize)
            {
                int chunkSize = allocSize - _front;
                Array.Copy(_buffer, _front, dest.Array, 0, chunkSize);
                Array.Copy(_buffer, 0, dest.Array, chunkSize, size - chunkSize);
            }
            else
            {
                Array.Copy(_buffer, _front, dest.Array, 0, size);
            }
            
            _front = (_front + size) % allocSize;
            return size;
        }

        public Int32 Peek(out ArraySegment<Byte> dest, Int32 size)
        {
            if (size > GetUseSize())
            {
                dest = null;
                return -1;
            }

            byte[] destArray = new byte[size];
            
            if (_front + size > bufferSize)
            {
                int chunkSize = allocSize - _front;
                Array.Copy(_buffer, _front, destArray, 0, chunkSize);
                Array.Copy(_buffer, 0, destArray, chunkSize, size - chunkSize);
            }
            else
            {
                Array.Copy(_buffer, _front, destArray, 0, size);
            }

            dest = new ArraySegment<byte>(destArray);
            
            return size;
        }
        
        public List<ArraySegment<Byte>> GetUseSegments()
        {
            List<ArraySegment<Byte>> wsabufs = new List<ArraySegment<byte>>();
            Int32 directSize = DirectDequeueSize();
            wsabufs.Add(new ArraySegment<Byte>(_buffer, _front, directSize));
            Int32 remainSize = GetUseSize() - directSize;
            if (directSize > 0)
            {
                wsabufs.Add(new ArraySegment<Byte>(_buffer, 0, remainSize));
            }
            return wsabufs;
        }
        
        public List<ArraySegment<Byte>> GetFreeSegments()
        {
            List<ArraySegment<Byte>> wsabufs = new List<ArraySegment<byte>>();
            Int32 directSize = DirectEnqueueSize();
            wsabufs.Add(new ArraySegment<Byte>(_buffer, _rear, directSize));

            Int32 remainSize = GetFreeSize() -directSize;
            if (directSize > 0)
            {
                wsabufs.Add(new ArraySegment<Byte>(_buffer, 0, remainSize));
            }
            return wsabufs;
        }
    }
}